daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala

```diff
@@ -92,12 +92,12 @@ class DFDLSchemaFileLoadErrorHandler(schemaFileLocation: SchemaFileLocation)
   }
 
   /**
-   * Called on a fatal exception. The parser/validator throws the exception after
-   * this call returns.
+   * Called on a fatal exception and throws the exception
    */
-  def fatalError(exception: SAXParseException) = error(
-    exception,
-  ) // same as non-fatal exception.
+  def fatalError(exception: SAXParseException) = {
+    loaderErrors_ :+= exception
+    throw exception
```

Comment:
Reading the javadoc for fatalError, it seems like we shouldn't need to throw an exception here, and that the parser should throw the exception after calling fatalError if it is unable to continue.

If this is needed to workaround a bug in Scala-xml that's fine, but it should be documented why it's needed. Maybe scala-xml expects fatalError implementations to throw the exceptions?

Reply:
I was reading the SE8 documentation for ErrorHandler and it has this paragraph:

>For XML processing errors, a SAX driver must use this interface in preference to throwing an exception: it is up to the application to decide whether to throw an exception for different types of errors and warnings. Note, however, that there is no requirement that the parser continue to report additional errors after a call to [fatalError](https://docs.oracle.com/javase/8/docs/api/org/xml/sax/ErrorHandler.html#fatalError-org.xml.sax.SAXParseException-). In other words, a SAX driver class may throw an exception after reporting any fatalError. Also parsers may throw appropriate exceptions for non-XML errors. For example, [XMLReader.parse()](https://docs.oracle.com/javase/8/docs/api/org/xml/sax/XMLReader.html#parse-org.xml.sax.InputSource-) would throw an IOException for errors accessing entities or the document.

From that I concluded that if you want to stop execution on a fatal error, an exception should be thrown in `fatalError` since it said the driver *may* throw an exception. I don't think the driver is required to throw an exception in this case and there is flexibility for it to throw one or not.

Reply:
> it up to the application to decide whether to throw an exception for different types of errors and warnings. 

This makes me think that we (the application) shouldn't need to throw an exception to get the behavior we need from Xerces (the driver). The behavior we want to just add warnings/errors/fatalErrots to a list and then to let the driver continue gathering errors if it can do so.

And no where does it say the fatalError implementation *must* throw the exception. If that's actually needed because of a bug in Xerces/Scala XML/whatever is calling fatalError requires that behavior, we should document that.

> In other words, a SAX driver class may throw an exception after reporting any fatalError.

This is what I would expect. Xerces calls fatal error, we handle it and add the error to our list, and then Xerces throws the exception if can't make progress OR continues calling error/fatalError more if it can. By throwing an exception, we are telling Xerces to stop even if it could continue, which I don't think is what we want--we want as many errors from Xerces as possible.

Reply:
This is needed to workaround what appears to be a bug in scala-xml and its trying to stop it because it is getting caught in that loop. Yes, I wouldn't expect that behavior and a comment can be added to explain what is going on. I was just saying that the driver may continue as much as it wants and doesn't have to throw an exception when the error handler is installed. 

As an alternative, we could always keep around the last error seen. If the next error is exactly the same as the last error, no progress is being made, and we could throw the exception at that point. That way we can get as many errors as possible without triggering the infinite loop.

Reply:
> As an alternative, we could always keep around the last error seen...

It's probably not worth the added complexity, I think just commenting about why we have the throw an limitations (e.g. maybe don't get as many errors as possible) is fine.

Also, I grepped through the scala-xml code and couldn't find any calls to `fatalError`. I wonder if under the hood it's just using Xerces or some Java built-in XML parser, and the issue is actually in that parser and not in scala-xml. Can you output a stack trace? It would be nice to open a ticket with whatever is doing this and see if it could be fixed.